src/mlpack/methods/ann/layer/policy_impl.hpp

```diff
@@ -0,0 +1,63 @@
+/**
+ * @file policy_impl.hpp
+ * @author Shangtong Zhang
+ *
+ * Implementation of the Policy class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_POLICY_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_POLICY_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "policy.hpp"
+
+namespace mlpack {
+namespace ann {
+
+template<typename InputDataType, typename OutputDataType>
+Policy<InputDataType, OutputDataType>::Policy(
+    double entropyRegularizationWeight) :
+    entropyRegularizationWeight(entropyRegularizationWeight)
+{ /* Nothing to do here. */ }
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void Policy<InputDataType, OutputDataType>::Forward(
+    const InputType &&input, OutputType &&output)
+{
+  arma::mat maxInput = arma::repmat(arma::max(input), input.n_rows, 1);
+  arma::mat expInput = arma::exp(input - maxInput);
+  arma::mat dense = arma::repmat(arma::sum(expInput), expInput.n_rows, 1);
+  output = expInput / dense;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Policy<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT> &&prob, arma::Mat<eT> &&advantage, arma::Mat<eT> &&g)
+{
+  arma::mat adv = arma::repmat(arma::sum(advantage), advantage.n_rows, 1);
+  arma::mat gradientPolicy = advantage - adv % prob;
+  arma::mat logProb = arma::log(prob);
```

Comment:
We should use ``trunc_log`` or add a small probability factor to prevent NaNs. Also did you wrote a test for the ``Policy`` layer, without using the Reinforce layer?

Reply:
Sure. https://github.com/ShangtongZhang/DeepRL/blob/mlpack/async_worker/actor_critic.py
And the definition of actor network and critic network is here https://github.com/ShangtongZhang/DeepRL/blob/mlpack/main.py, the entrance function is `a3c_cart_pole`

Reply:
Okay, thanks, let me take a look.

Reply:
Since you used the same hyperparameter, I think we don't have to test different parameter sets.